src/com/facebook/buck/go/GoBinary.java

```diff
@@ -0,0 +1,142 @@
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildableProperties;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.step.Step;
+import com.facebook.buck.step.fs.MakeCleanDirectoryStep;
+import com.facebook.buck.step.fs.MkdirStep;
+import com.facebook.buck.step.fs.SymlinkTreeStep;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class GoBinary extends AbstractBuildRule implements BinaryBuildRule {
+
+  @AddToRuleKey
+  private final Tool linker;
+  @AddToRuleKey
+  private final ImmutableList<String> linkerFlags;
+  @AddToRuleKey
+  private final Linker cxxLinker;
+
+  private final GoLinkable lib;
+
+  private final Path scratchDir;
+  private final Path output;
+
+  public GoBinary(
+      BuildRuleParams params,
+      SourcePathResolver resolver,
+      Linker cxxLinker,
+      GoLinkable lib,
+      Tool linker,
+      ImmutableList<String> linkerFlags) {
+    super(params, resolver);
+    this.cxxLinker = cxxLinker;
+    this.scratchDir = BuildTargets.getScratchPath(getBuildTarget(), "__%s__link");
+    this.linker = linker;
+    this.lib = lib;
+    this.output = BuildTargets.getGenPath(
+        params.getBuildTarget(), "%s/" + params.getBuildTarget().getShortName());
+    this.linkerFlags = linkerFlags;
+  }
+
+  @Override
+  public Tool getExecutableCommand() {
+    return new CommandTool.Builder()
+        .addArg(new BuildTargetSourcePath(getBuildTarget()))
+        .build();
+  }
+
+  @Override
+  public BuildableProperties getProperties() {
+    return new BuildableProperties(BuildableProperties.Kind.PACKAGING);
+  }
+
+  @Override
+  public ImmutableList<Step> getBuildSteps(
+      BuildContext context, BuildableContext buildableContext) {
+    final ImmutableMap.Builder<Path, Path> symlinkTree = ImmutableMap.builder();
+    AbstractBreadthFirstTraversal<GoLinkable> visitor = new AbstractBreadthFirstTraversal<GoLinkable>(lib) {
+      @Override
+      public ImmutableSet<GoLinkable> visit(GoLinkable goLinkable) {
+        Path ruleOutput = goLinkable.getPathToOutput();
+        Preconditions.checkNotNull(ruleOutput);
+
+        String extension = Files.getFileExtension(ruleOutput.toString());
+        Path symlinkPath = Paths.get(goLinkable.getPackageName().toString() +
+                (extension.equals("") ? "" : "." + extension));
+        symlinkTree.put(symlinkPath, ruleOutput);
+
+        return FluentIterable.from(goLinkable.getDeps()).transform(
+            new Function<BuildRule, GoLinkable>() {
+              @Override
+              public GoLinkable apply(BuildRule b) {
+                return (GoLinkable) b;
```

Comment:
Throw a `HumanReadableException` if `b` isn't a `GoLinkable`, so the user sees something like "go_binary deps must contain only go_binary or go_library targets" instead of "class cast exception 20 lines of stack trace" if they try to depend on something that isn't `GoLinkable`.